---
title: "<small>Towards</small> a reproducible and reusable publication and analysis workflow"
author:
- Rad Suchecki
- Nathan Watson-Haigh
- Stuart Stephen
- Alex Whan
#subtitle: "Subtitle"
date: "`r format(Sys.time(), '%A, %d %B, %Y')`"
#tags: [nothing, nothingness]
output:
  #pdf_document:
    #default
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    theme: yeti
    highlight: tango
  beamer_presentation:
    highlight: default #specifies the syntax highlighting style. Supported styles include “default”, “tango”, “pygments”, “kate”, “monochrome”, “espresso”, “zenburn”, and “haddock” (specify null to prevent syntax highlighting)
    toc: true
    slide_level: 2
  revealjs::revealjs_presentation:
    theme: beige #“default”, “simple”, “sky”, “beige”, “serif”, “solarized”, “blood”, “moon”, “night”, “black”, “league” or “white”
    highlight: default # “default”, “tango”, “pygments”, “kate”, “monochrome”, “espresso”, “zenburn”, and “haddock”. Pass null to prevent syntax highlighting.
    center: false #specifies whether you want to vertically center content on slides (this defaults to false).
    transition: slide #"default”, “fade”, “slide”, “convex”, “concave”, “zoom” or “none”.
    incremental: false
    css: reveal.css
    slide_level: 1 #default=2
    self_contained: true
    reveal_options:
      slideNumber: true
      margin: 0.1
      width: 1280
      height: 800
      preview_links: true
  slidy_presentation:
    font_adjustment: -1
  ioslides_presentation:
    widescreen: true
    transition: faster

---



```{r include = FALSE}
# <mark>Towards</mark> a reproducible and reusable publication and analysis workflow
#knitr::opts_chunk$set(echo = FALSE) #resets defaults
#library(viridis) #load required libs here
```




## BioKanga {incremental=false}

- A suite of bioinformatics tools developed at CSIRO.
- [https://github.com/csiro-crop-informatics/biokanga](https://github.com/csiro-crop-informatics/biokanga)



- Aim:
  * evaluation of BioKanga's sequence alignment module vs state-of-the-art tools
  * turn-key reproducibility


----

## Modular workflow
- we developed a modular workflow using ~~two popular automation frameworks, Snakemake and~~ [Nextflow](https://www.nextflow.io/)

- Individual tasks are executed in Cloud or HPC environment using dedicated, ~~lightweight~~ (Docker/Singularity) containers.
- Raw data is acquired from a public sources, processed and analysed.
- Final document including dynamically generated figures and tables is compiled from R Markdown <strike> or LaTeX/knitr </strike>.

----

## [Nextflow](https://www.nextflow.io/) 

A Workflow management system which promises 

  * Portability
  * Scalability
  * Reproducibility

Features 

  * Separation of pipeline logic from compute environment definition(s)
  * Syntax is a superset of Groovy, but
  * Polyglot
    * mixing scripting languages
  * Multiple _executors_
    * SGE, LSF, SLURM, PBS/Torque, NQSII HTCondor, Ignite, Kubernets, AWS Batch

[https://www.nature.com/articles/nbt.3820/](https://www.nature.com/articles/nbt.3820)


----

## Pipeline flow

```{bash, echo = FALSE}
dot -Tpng -odoc/flowchart.png doc/flowchart.dot
dot -Tpng -Gsize=9,7\! -Gdpi=100 -odoc/flowchart_small.png doc/flowchart.dot
```

[![Flowchart](doc/flowchart_small.png)](doc/flowchart.png)

----

## Pipeline syntax

* Primarily definitions of processes, each with well defined input and output channels

```{bash, comment=''}
cat main.nf
```

----

## Compute profiles 

* __Implemented:__
  - `standard` i.e. local/interactive assuming all software available
  - `slurm`
  - `modules`
  - `docker`
  - `singularity`
  - `ec2`

Some of the profiles are intended to be used in concert, e.g. 

```{bash, eval = FALSE}
nextflow run main.nf -profile slurm,modules
```

* __Upcoming:__ `k8s`, `awsbatch`

----


## Modules

* Required modules are defined separately and explicitly for each process

```{bash, comment=''}
cat conf/modules.config
```

* Unfortunately, these are compute environment specific
* On the up-side - updating these should be all that is required to run the pipeline on a different SLURM cluster

----

## Containers

* Required containers are defined separately and explicitly for each process
* Available on docker hub
  * Used both by `docker` and `singularity`
  * Always specify version tag (avoid `':latest'`)

```{bash, comment=''}
cat conf/containers.config
```

----

## Generating reports/publications/slides from R Markdown 

![Tool-chain](https://d33wubrfki0l68.cloudfront.net/61d189fd9cdf955058415d3e1b28dd60e1bd7c9b/9791d/images/rmarkdownflow.png)

### Some of the supported output formats:

* HTML* documents and slide-shows using 
  * Slidy, ioslides, **reveal.js**
* PDF
  * LaTeX
  * *LaTeX Beamer slides*
* MS Word, ODT, RTF 
* Markdown, and Github flavored Markdown documents

----

## This presentation

* Dependencies
  * R, Pandoc, `rmarkdown`, `revealjs`
* Use R, RStudio or a simple `RScript` from the command line

```{bash, eval = FALSE}
./render.R show.Rmd
```

```{bash, comment='', echo = FALSE}
head -15 render.R
```

----

## Report generation

* results of tested tools are compared to ground-trouth (original locations of simulated reads)
* 



----

## Further Work

* More systematic evaluation of BioKanga alignment module vs other tools
  * At default settings
  * Exploration of parameter space, optimisation
* Refinement of document generation, as R Markdown rendering not as flexible as hoped
  * Generation of ouptut in different formats somewhat limited
  * For more complex content where pure (R) Markdown is insuffcient have to decide on either 
    * (R) Markdown with LaTeX 
    * (R) Markdown with HTML
* Evaluation of Nextflow against Snakemake



